rabbitmq/src/main/java/com/rabbitmq/jepsen/Utils.java

```diff
@@ -0,0 +1,299 @@
+        public void setup() throws Exception {
+            try (Channel ch = connection.createChannel()) {
+                ch.queueDeclare(QUEUE, true, false, false, Collections.singletonMap("x-queue-type", "quorum"));
+                Thread.sleep(1000);
+                ch.queuePurge(QUEUE);
+            }
+        }
+
+        public boolean enqueue(Object value, Number publishConfirmTimeout) throws Exception {
+            initializeIfNecessary();
+            publishingChannel.basicPublish("", QUEUE,
+                    true,
+                    new AMQP.BasicProperties.Builder().deliveryMode(2).build(),
+                    value.toString().getBytes()
+            );
+            return publishingChannel.waitForConfirms(publishConfirmTimeout.intValue());
+        }
+
+        public void close() throws Exception {
+            connection.close(5000);
+        }
+
+    }
+
+    static class AsynchronousConsumerClient extends AbstractClient {
+
+        private final Queue<Delivery> enqueued = new ConcurrentLinkedDeque<>();
+        private final CountDownLatch cancelOkLatch = new CountDownLatch(1);
+
+        private volatile String consumerTag;
+
+        public AsynchronousConsumerClient(String host) throws Exception {
+            super(host);
+        }
+
+        public Integer dequeue() throws Exception {
+            initializeIfNecessary();
+            Delivery delivery = enqueued.poll();
+            if (delivery == null) {
+                return null;
+            } else {
+                consumingChannel.basicAck(delivery.getEnvelope().getDeliveryTag(), false);
+                return Integer.valueOf(new String(delivery.getBody()));
+            }
+        }
+
+        public IPersistentVector drain() throws Exception {
+            Collection<Integer> values = new CopyOnWriteArrayList<>();
+            if (isInitialized()) {
+                consumingChannel.basicCancel(consumerTag);
+                // doing our best to wait for the long-running consumer to be stopped
+                // to drain the local queue safely
+                cancelOkLatch.await(1, TimeUnit.SECONDS);
+                Integer value = dequeue();
+                while (value != null) {
+                    values.add(value);
+                    value = dequeue();
+                }
+                try {
+                    consumingChannel.close();
+                } catch (Exception e) {
+                    // ignoring, we want to drain
+                }
+            }
+
+            Channel ch = connection.createChannel();
+            ch.basicConsume(QUEUE, false, (ctag, delivery) -> {
+                values.add(Integer.valueOf(new String(delivery.getBody())));
+                ch.basicAck(delivery.getEnvelope().getDeliveryTag(), false);
+            }, ctag -> {
+            });
+
+            consumingChannel = connection.createChannel();
+            while (consumingChannel.queueDeclarePassive(QUEUE).getMessageCount() != 0) {
```

Comment:
I think there is a potential problem with this code. passive queue declares return the number of ready messages _not_ the total number of messages inside the queue including those awaiting acks. We could have a potential race condition here that the values are returned before the draining consumer acks any in flight messages resulting in a delivery not being recorded. Maybe moving to polling drains would be more reliable?

Reply:
We can use Hop to query the total number of messages.